File: 0200-number-of-islands/0200-number-of-islands.js
/**
 * @param {character[][]} grid
 * @return {number}
 */
var numIslands = function(grid) {
    function fillIsland(row,col){
        if(row<0 || row===grid.length) return;
        if(col<0 || col===grid[0].length) return;
        if(grid[row][col] === '0') return;
        grid[row][col] = '0';
        fillIsland(row+1,col);
        fillIsland(row,col+1);
        fillIsland(row-1,col);
        fillIsland(row,col-1);
    }
    let count = 0;
    for(let row = 0; row < grid.length; row++){
        for(let col = 0; col < grid[0].length; col++){
            if(grid[row][col] === '1'){
                count++;
                fillIsland(row,col);
            }
        }
    }
    return count;
};

File: 0200-number-of-islands/0200-number-of-islands.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        M, N = len(grid), len(grid[0])
        def fill(r,c):
            if r<0 or c<0 or r==M or c==N:
                return
            if grid[r][c] == '0':
                return
            grid[r][c] = '0'
            fill(r-1,c)
            fill(r,c+1)
            fill(r+1,c)
            fill(r,c-1)
        
        islands = 0
        for r in range(M):
            for c in range(N):
                if grid[r][c]=='1':
                    islands += 1
                    fill(r,c)
        return islands